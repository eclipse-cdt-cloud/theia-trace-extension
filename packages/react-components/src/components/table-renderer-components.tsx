import { faAngleDown, faSearch, faSpinner, faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ICellRendererParams, IFloatingFilterParams } from 'ag-grid-community';
import debounce from 'lodash.debounce';
import * as React from 'react';

type CellRendererProps = ICellRendererParams & {
    filterModel: Map<string, string>;
    searchResultsColor: string;
};

type SearchFilterRendererProps = IFloatingFilterParams & {
    onFilterChange: (colName: string, filterValue: string) => void;
    onclickNext: () => void;
    colName: string;
};

interface SearchFilterRendererState {
    hasHovered: boolean;
    hasClicked: boolean;
}

export class LoadingRenderer extends React.Component<ICellRendererParams> {
    render(): JSX.Element {
        return <React.Fragment>
            {this.props.value ? this.props.value : <FontAwesomeIcon icon={faSpinner} />}
        </React.Fragment>;
    }
}

export class CellRenderer extends React.Component<CellRendererProps> {
    render(): JSX.Element {
        if (this.props.value === undefined) {
            return <React.Fragment><FontAwesomeIcon icon={faSpinner} /></React.Fragment>;
        }
        const currField = this.props.colDef?.field;
        const currFieldVal = this.props.value || '';
        const isMatched = this.props.data && this.props.data['isMatched'];
        let cellElement: JSX.Element;
        const searchTerm = (currField && this.props.filterModel.get(currField)) || '';
        if (this.props.filterModel.size > 0) {
            if (isMatched) {
                cellElement = <>
                    {currFieldVal.split(searchTerm).map((tag: string, index: number, array: string[]) =>
                        <span key={index.toString()}>
                            <>
                                {tag}
                            </>
                            <>
                                {
                                    index < array.length - 1 ?
                                        <span style={{ backgroundColor: this.props.searchResultsColor }}>{searchTerm}</span>
                                        : <></>
                                }
                            </>
                        </span>
                    )}
                </>;
            }
            else {
                cellElement = <span style={{ color: '#808080' }}> {currFieldVal} </span>;
            }
        } else {
            cellElement = <>{currFieldVal || ''}</>;
        }

        return <React.Fragment>
            {cellElement}
        </React.Fragment>;
    }
}

export class SearchFilterRenderer extends React.Component<SearchFilterRendererProps, SearchFilterRendererState> {

    private debouncedChangeHandler: (colName: string, inputVal: string) => void;

    constructor(props: SearchFilterRendererProps) {
        super(props);
        this.state = {
            hasHovered: false,
            hasClicked: false
        };

        this.debouncedChangeHandler = debounce((colName, inputVal) => {
            this.props.onFilterChange(colName, inputVal);
        }, 500);

        this.onInputBoxChanged = this.onInputBoxChanged.bind(this);
        this.onMouseEnterHandler = this.onMouseEnterHandler.bind(this);
        this.onMouseLeaveHandler = this.onMouseLeaveHandler.bind(this);
        this.onClickHandler = this.onClickHandler.bind(this);
        this.onDownClickHandler = this.onDownClickHandler.bind(this);
        this.onCloseClickHandler = this.onCloseClickHandler.bind(this);
        this.onKeyDownEvent = this.onKeyDownEvent.bind(this);

    }

    render(): JSX.Element {
        return (
            <div onMouseEnter={this.onMouseEnterHandler} onMouseLeave={this.onMouseLeaveHandler} onClick={this.onClickHandler}>
                {!this.state.hasClicked && !this.state.hasHovered &&
                    <FontAwesomeIcon style={{ marginLeft: '10px' }} icon={faSearch} />}
                {!this.state.hasClicked && this.state.hasHovered &&
                    <div style={{ cursor: 'pointer' }}>
                        <FontAwesomeIcon style={{ marginLeft: '10px' }} icon={faSearch} />
                        <span style={{ marginLeft: '10px', fontSize: '13px', width: '50px', overflow: 'hidden', textOverflow: 'ellipsis' }}>Search</span>
                    </div>}
                {this.state.hasClicked &&
                    <div>
                        <input type="text" autoFocus={true} onKeyDown={this.onKeyDownEvent} onInput={this.onInputBoxChanged} style={{ width: '50%', margin: '10px' }} />
                        <FontAwesomeIcon className='hoverClass' icon={faTimes} style={{ marginTop: '20px' }} onClick={this.onCloseClickHandler} />
                        <FontAwesomeIcon className='hoverClass' icon={faAngleDown} style={{ marginLeft: '10px', marginTop: '20px' }} onClick={this.onDownClickHandler} />
                    </div>}
            </div>
        );
    }

    private onKeyDownEvent(event: React.KeyboardEvent) {
        if (event.key === 'Enter') {
            this.props.onclickNext();
        } else if (event.key === 'Escape') {
            this.setState({
                hasClicked: false
            });
            this.props.onFilterChange(this.props.colName, '');
        }
        return;
    }

    private onInputBoxChanged(event: React.ChangeEvent<HTMLInputElement>) {
        this.debouncedChangeHandler(this.props.colName, event.target.value);
        return;
    }

    private onMouseEnterHandler() {
        this.setState({
            hasHovered: true
        });
        return;
    }

    private onMouseLeaveHandler() {
        this.setState({
            hasHovered: false
        });
        return;
    }

    private onClickHandler() {
        this.setState({
            hasClicked: true
        });
        return;
    }

    private onDownClickHandler() {
        this.props.onclickNext();
        return;
    }

    private onCloseClickHandler(event: React.MouseEvent) {
        this.setState({
            hasClicked: false
        });
        this.props.onFilterChange(this.props.colName, '');
        event.stopPropagation();
        return;
    }
}
