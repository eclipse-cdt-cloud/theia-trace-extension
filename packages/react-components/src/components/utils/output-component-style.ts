export interface OutputComponentStyle {
    width: number;
    chartWidth: number;
    // react-grid-layout - The library used for resizing components
    // inserts new React components during compilation, and the dimensions
    // it returns are strings (pixels).
    // Currently, the components are only height-resizable.
    height: number | string;
    naviBackgroundColor: number;
    chartBackgroundColor: number;
    cursorColor: number;
    lineColor: number;
    rowHeight: number;
}
