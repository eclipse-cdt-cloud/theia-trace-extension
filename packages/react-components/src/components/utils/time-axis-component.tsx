import * as React from 'react';
import { TimeGraphAxis } from 'timeline-chart/lib/layer/time-graph-axis';
import { TimeGraphAxisCursors } from 'timeline-chart/lib/layer/time-graph-axis-cursors';
import { ReactTimeGraphContainer } from './timegraph-container-component';
import { TimeGraphUnitController } from 'timeline-chart/lib/time-graph-unit-controller';

interface TimeAxisProps {
    unitController: TimeGraphUnitController;
    style: {
        width: number,
        chartBackgroundColor: number,
        cursorColor: number,
        lineColor: number,
        verticalAlign: string
    };
    addWidgetResizeHandler: (handler: () => void) => void;
    removeWidgetResizeHandler: (handler: () => void) => void;
}

export class TimeAxisComponent extends React.Component<TimeAxisProps> {
    render(): JSX.Element {
        return <ReactTimeGraphContainer
            id='timegraph-axis'
            options={{
                id: 'timegraph-axis',
                width: this.props.style.width,
                height: 30,
                backgroundColor: this.props.style.chartBackgroundColor,
                lineColor: this.props.style.lineColor,
                classNames: 'horizontal-canvas'
            }}
            addWidgetResizeHandler={this.props.addWidgetResizeHandler}
            removeWidgetResizeHandler={this.props.removeWidgetResizeHandler}
            unitController={this.props.unitController}
            layers={[this.getAxisLayer(), this.getAxisCursors()]} />;
    }

    protected getAxisLayer(): TimeGraphAxis {
        const timeAxisLayer = new TimeGraphAxis('timeGraphAxis', {
            color: this.props.style.chartBackgroundColor,
            lineColor: this.props.style.lineColor,
            verticalAlign: this.props.style.verticalAlign
        });
        return timeAxisLayer;
    }

    protected getAxisCursors(): TimeGraphAxisCursors {
        return new TimeGraphAxisCursors('timeGraphAxisCursors', { color: this.props.style.cursorColor });
    }
}
