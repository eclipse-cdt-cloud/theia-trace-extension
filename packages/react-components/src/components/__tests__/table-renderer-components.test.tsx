import * as React from 'react';
import { cleanup, render, fireEvent, screen, act } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { TableOutputComponent } from '../table-output-component';
import { CellRenderer, LoadingRenderer, SearchFilterRenderer } from '../table-renderer-components';
import { AbstractOutputProps } from '../abstract-output-component';
import { TimeGraphUnitController } from 'timeline-chart/lib/time-graph-unit-controller';
import { TimeRange } from 'traceviewer-base/lib/utils/time-range';
import { TspClient } from 'tsp-typescript-client/lib/protocol/tsp-client';
import { ColDef, Column, ColumnApi, GridApi, IRowModel, RowNode } from 'ag-grid-community';

describe('<TableOutputComponent />', () => {

    let tableComponent: any;
    const ref = (el: TableOutputComponent | undefined | null): void => {
        tableComponent = el;
    };

    beforeEach(() => {
        tableComponent = null;
    });

    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    });

    test('Renders AG-Grid table with provided props & state', async () => {
        const tableOutputComponentProps: AbstractOutputProps = {
            tooltipComponent: null,
            style: {
                width: 0,
                height: 0,
                rowHeight: 0,
                naviBackgroundColor: 0,
                chartBackgroundColor: 0,
                cursorColor: 0,
                lineColor: 0,
                componentLeft: 0,
                chartOffset: 0
            },
            tspClient: new TspClient('testURL'),
            traceId: '0',
            outputDescriptor: {
                id: '0',
                name: '',
                description: '',
                type: '',
                queryParameters: new Map<string, any>(),
                start: BigInt(0),
                end: BigInt(0),
                final: true,
                compatibleProviders: []            
            },
            markerCategories: undefined,
            markerSetId: '0',
            range: new TimeRange(BigInt(0), BigInt(0)),
            nbEvents: 0,
            viewRange: new TimeRange(BigInt(0), BigInt(0)),
            selectionRange: undefined,
            onOutputRemove: () => 0,
            unitController: new TimeGraphUnitController(BigInt(0), { start: BigInt(0), end: BigInt(0)}),
            backgroundTheme: 'light',
            tooltipXYComponent: null,
            outputWidth: 0
        };

        const tableOutputComponentState = [
            {
                "headerName":"Trace",
                "field":"0",
                "width":0,
                "resizable":true,
                "cellRenderer":"cellRenderer",
                "cellRendererParams":{"filterModel":{},"searchResultsColor":"#cccc00"},
                "suppressMenu":true,
                "filter":"agTextColumnFilter",
                "floatingFilter":true,
                "floatingFilterComponent":"searchFilterRenderer",
                "floatingFilterComponentParams":{"suppressFilterButton":true,"colName":"0"},
                "icons":{"filter":""},
                "tooltipField":"0"
            }
        ];

        const { container } = render(<TableOutputComponent {...tableOutputComponentProps} ref={ref} />);
        expect(tableComponent).toBeTruthy();
        expect(tableComponent instanceof TableOutputComponent).toBe(true);
        const table: TableOutputComponent = tableComponent;
        act(() => {
            // fire events that update state
            table.setState(
                {tableColumns: tableOutputComponentState}
            );
          });

        // Renders with provided props
        expect(table.state.tableColumns).toEqual(tableOutputComponentState);
        expect(table.props.backgroundTheme).toEqual('light');
        expect(table.props.tooltipComponent).toEqual(null);
        expect(table.props.style).toEqual({
            width: 0,
            height: 0,
            rowHeight: 0,
            naviBackgroundColor: 0,
            chartBackgroundColor: 0,
            cursorColor: 0,
            lineColor: 0,
            componentLeft: 0,
            chartOffset: 0
        });
        expect(container).toMatchSnapshot();
    });

    const mockOnFilterChange = jest.fn();
    const mockOnClickNext = jest.fn();
    const mockOnClickPrevious = jest.fn();
    const mockParentilterInstance = jest.fn();
    const mockCurrentParentModel = jest.fn();
    const mockFilterChangedCallback = jest.fn();
    const mockFilterModifiedCallback = jest.fn();
    const mockValueGetter = jest.fn();
    const mockDoesRowPassOtherFilter = jest.fn();
    const mockShowParentFilter = jest.fn();

    test('Empty search filter renderer', () => {
        const searchFilter = create(<SearchFilterRenderer 
            colName={'jest Test'}
            onFilterChange={mockOnFilterChange}
            onclickNext={mockOnClickNext}
            onclickPrevious={mockOnClickPrevious} 
            column={new Column({} as ColDef, {} as ColDef, 'jest Test', true)} 
            filterParams={
                {
                    api: new GridApi(),
                    column: new Column({} as ColDef, {} as ColDef, 'jest Test', true),
                    colDef: {} as ColDef,
                    columnApi: new ColumnApi(),
                    rowModel: '' as unknown as IRowModel,
                    filterChangedCallback: mockFilterChangedCallback,
                    filterModifiedCallback: mockFilterModifiedCallback,
                    valueGetter: mockValueGetter,
                    doesRowPassOtherFilter: mockDoesRowPassOtherFilter,
                    context: '',
                }
            } 
            currentParentModel={mockCurrentParentModel} 
            parentFilterInstance={mockParentilterInstance} 
            suppressFilterButton={false} 
            api={new GridApi()} 
            columnApi={new ColumnApi()}
            showParentFilter={mockShowParentFilter}
            context={''}
            filterModel= {new Map<string,string>()}
        />).toJSON();
        expect(searchFilter).toMatchSnapshot();
    });

    test('Search filter renderer key interactions', () => {
        render(<SearchFilterRenderer 
            colName={'jest Test'}
            onFilterChange={mockOnFilterChange}
            onclickNext={mockOnClickNext}
            onclickPrevious={mockOnClickPrevious} 
            column={new Column({} as ColDef, {} as ColDef, 'jest Test', true)} 
            filterParams={
                {
                    api: new GridApi(),
                    column: new Column({} as ColDef, {} as ColDef, 'jest Test', true),
                    colDef: {} as ColDef,
                    columnApi: new ColumnApi(),
                    rowModel: '' as unknown as IRowModel,
                    filterChangedCallback: mockFilterChangedCallback,
                    filterModifiedCallback: mockFilterModifiedCallback,
                    valueGetter: mockValueGetter,
                    doesRowPassOtherFilter: mockDoesRowPassOtherFilter,
                    context: '',
                }
            } 
            currentParentModel={mockCurrentParentModel} 
            parentFilterInstance={mockParentilterInstance} 
            suppressFilterButton={false} 
            api={new GridApi()}
            columnApi={new ColumnApi()}
            showParentFilter={mockShowParentFilter}
            context={''}
            filterModel={new Map<string,string>()}
        />);

        const parentDiv = screen.getByTestId('search-filter-element-parent');
        fireEvent.click(parentDiv);
        const input = screen.getByTestId('search-filter-element-input');
        fireEvent.keyDown(input, {key: 'Enter', code: 'Enter', charCode: 13});
        expect(mockOnClickNext).toHaveBeenCalledTimes(1);
        fireEvent.keyDown(input, {key: 'Escape', code: 'Escape', charCode: 27});
        expect(mockOnFilterChange).toHaveBeenCalledTimes(1);
    });

    const cellRendererProps: any = {
        value: 'test cell',
        valueFormatted: 'test cell',
        getValue: function () {
            return 'test cell';
        },
        setValue: function (val: any): void {
            const value = val;
        },
        formatValue: function (value: any) {
            return value;
        },
        data: {},
        node: {} as RowNode,
        colDef: {} as ColDef,
        column: new Column({} as ColDef, {} as ColDef, 'jest Test', true),
        rowIndex: 0,
        api: new GridApi(),
        columnApi: new ColumnApi(),
        context: '',
        refreshCell: () => {},
        $scope: '',
        eGridCell: document.createElement('div'),
        eParentOfValue: document.createElement('div'),
        addRenderedRowListener: function (eventType: string, listener: Function): void {
            throw new Error('Function not implemented.');
        },
        searchResultsColor: '#FFFF00',
        filterModel: new Map<string,string>()
    }

    test('Cell renderer', () => {
        const cellRenderer = create(<CellRenderer {...cellRendererProps}/>).toJSON();
        expect(cellRenderer).toMatchSnapshot();
    });

    test('Cell renderer with search selection', () => {
        cellRendererProps.colDef.field = 'testField';
        cellRendererProps.filterModel.set('testField', 'test');
        cellRendererProps.data = {'isMatched' : true};

        const cellRenderer = create(<CellRenderer {...cellRendererProps}/>).toJSON();
        expect(cellRenderer).toMatchSnapshot();
    });

    test('Loading renderer not loading', () => {
        delete cellRendererProps.filterModel;
        delete cellRendererProps.searchResultsColor;

        const loadingRenderer = create(<LoadingRenderer {...cellRendererProps}/>).toJSON();
        expect(loadingRenderer).toMatchSnapshot();
    });

    test('Loading renderer in loading', () => {
        cellRendererProps.value = undefined;

        const loadingRenderer = create(<LoadingRenderer {...cellRendererProps}/>).toJSON();
        expect(loadingRenderer).toMatchSnapshot();
    });

})
