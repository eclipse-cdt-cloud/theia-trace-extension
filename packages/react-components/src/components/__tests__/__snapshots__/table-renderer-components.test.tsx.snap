// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableOutputComponent /> Cell renderer 1`] = `"test cell"`;

exports[`<TableOutputComponent /> Cell renderer with search selection 1`] = `
Array [
  <span>
    <span
      style={
        Object {
          "backgroundColor": "#FFFF00",
        }
      }
    >
      test
    </span>
  </span>,
  <span>
     cell
  </span>,
]
`;

exports[`<TableOutputComponent /> Empty search filter renderer 1`] = `
<div
  data-testid="search-filter-element-parent"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-search fa-w-16 "
    data-icon="search"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "marginLeft": "10px",
      }
    }
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`<TableOutputComponent /> Loading renderer in loading 1`] = `
<svg
  aria-hidden="true"
  className="svg-inline--fa fa-spinner fa-w-16 "
  data-icon="spinner"
  data-prefix="fas"
  focusable="false"
  role="img"
  style={Object {}}
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
    fill="currentColor"
    style={Object {}}
  />
</svg>
`;

exports[`<TableOutputComponent /> Loading renderer not loading 1`] = `"test cell"`;

exports[`<TableOutputComponent /> Renders AG-Grid table with provided props & state 1`] = `
<div>
  <div
    class="output-container undefined"
    data-for="tooltip-component"
    data-tip=""
    id="00"
    style="width: 0px; height: 0px;"
    tabindex="-1"
  >
    <div
      class="widget-handle-with-options"
      id="00handle"
      style="width: 30px; height: 0px;"
    >
      <button
        class="remove-component-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="options-menu-container"
      >
        <button
          class="options-menu-button"
          title="Show View Options"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bars fa-w-14 "
            data-icon="bars"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="title-bar-label"
        title=""
      >
        <span />
        <span
          class="hidden-ellipsis"
        >
          ...
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-spinner fa-w-16 fa-spin "
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          id="00handleSpinner"
          role="img"
          style="margin-top: 5px; visibility: hidden;"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="main-output-container"
      style="width: -30px; height: 0px;"
    >
      <div
        class="ag-theme-balham"
        id="00focusContainer"
        style="height: 0px; width: 0px; display: flex; flex-direction: column;"
        tabindex="-1"
      >
        <div
          style="height: 100%;"
        >
          <!-- AG Grid -->
          <div
            class="ag-root-wrapper ag-ltr ag-layout-normal"
            role="presentation"
          >
            <div
              class="ag-root-wrapper-body ag-focus-managed ag-layout-normal"
              role="presentation"
            >
              <div
                class="ag-tab-guard ag-tab-guard-top"
                role="presentation"
                tabindex="0"
              />
              <!-- AG Grid Body -->
              <div
                aria-colcount="0"
                aria-rowcount="-1"
                class="ag-root ag-unselectable ag-layout-normal"
                role="grid"
              >
                <div
                  class="ag-header ag-pivot-off"
                  role="presentation"
                  style="height: 33px; min-height: 33px;"
                >
                  <div
                    class="ag-pinned-left-header ag-hidden"
                    role="presentation"
                    style="width: 0px; min-width: 0px; max-width: 0px;"
                  >
                    <div
                      aria-rowindex="1"
                      class="ag-header-row ag-header-row-column"
                      role="row"
                      style="height: 32px; top: 0px; width: 0px;"
                    />
                  </div>
                  <div
                    class="ag-header-viewport "
                    role="presentation"
                  >
                    <div
                      class="ag-header-container"
                      role="rowgroup"
                      style="width: 0px;"
                    >
                      <div
                        aria-rowindex="1"
                        class="ag-header-row ag-header-row-column"
                        role="row"
                        style="height: 32px; top: 0px; width: 0px;"
                      />
                    </div>
                  </div>
                  <div
                    class="ag-pinned-right-header ag-hidden"
                    role="presentation"
                    style="width: 0px; min-width: 0px; max-width: 0px;"
                  >
                    <div
                      aria-rowindex="1"
                      class="ag-header-row ag-header-row-column"
                      role="row"
                      style="height: 32px; top: 0px; width: 0px;"
                    />
                  </div>
                </div>
                <!-- AG Pinned Top -->
                <div
                  class="ag-floating-top"
                  role="presentation"
                  style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
                >
                  <!-- AG Row Container topLeft -->
                  <div
                    class="ag-pinned-left-floating-top ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container topCenter -->
                  <div
                    class="ag-floating-top-viewport"
                    role="presentation"
                  >
                    <div
                      class="ag-floating-top-container"
                      role="presentation"
                      style="width: 0px;"
                    />
                  </div>
                  <!-- AG Row Container topRight -->
                  <div
                    class="ag-pinned-right-floating-top ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container topFullWidth -->
                  <div
                    class="ag-floating-top-full-width-container"
                    role="presentation"
                  />
                </div>
                <!-- AG Middle -->
                <div
                  class="ag-body-viewport ag-row-no-animation ag-layout-normal"
                  role="presentation"
                >
                  <!-- AG Row Container left -->
                  <div
                    class="ag-pinned-left-cols-container ag-hidden"
                    role="presentation"
                    style="height: 1px;"
                  />
                  <!-- AG Row Container center -->
                  <div
                    class="ag-center-cols-clipper"
                    role="presentation"
                    style="height: 1px;"
                  >
                    <div
                      class="ag-center-cols-viewport"
                      role="presentation"
                      style="height: 100%;"
                    >
                      <div
                        class="ag-center-cols-container"
                        role="presentation"
                        style="height: 1px; width: 0px;"
                      />
                    </div>
                  </div>
                  <!-- AG Row Container right -->
                  <div
                    class="ag-pinned-right-cols-container ag-hidden"
                    role="presentation"
                    style="height: 1px;"
                  />
                  <!-- AG Row Container fullWidth -->
                  <div
                    class="ag-full-width-container"
                    role="presentation"
                    style="height: 1px;"
                  />
                </div>
                <!-- AG Sticky Top -->
                <div
                  class="ag-sticky-top"
                  role="presentation"
                  style="height: 0px; top: 33px; width: 100%;"
                >
                  <!-- AG Row Container stickyTopLeft -->
                  <div
                    class="ag-pinned-left-sticky-top ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container stickyTopCenter -->
                  <div
                    class="ag-sticky-top-viewport"
                    role="presentation"
                  >
                    <div
                      class="ag-sticky-top-container"
                      role="presentation"
                      style="width: 0px;"
                    />
                  </div>
                  <!-- AG Row Container stickyTopRight -->
                  <div
                    class="ag-pinned-right-sticky-top ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container stickyTopFullWidth -->
                  <div
                    class="ag-sticky-top-full-width-container"
                    role="presentation"
                  />
                </div>
                <!-- AG Pinned Bottom -->
                <div
                  class="ag-floating-bottom"
                  role="presentation"
                  style="height: 0px; min-height: 0; display: none; overflow-y: hidden;"
                >
                  <!-- AG Row Container bottomLeft -->
                  <div
                    class="ag-pinned-left-floating-bottom ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container bottomCenter -->
                  <div
                    class="ag-floating-bottom-viewport"
                    role="presentation"
                  >
                    <div
                      class="ag-floating-bottom-container"
                      role="presentation"
                      style="width: 0px;"
                    />
                  </div>
                  <!-- AG Row Container bottomRight -->
                  <div
                    class="ag-pinned-right-floating-bottom ag-hidden"
                    role="presentation"
                  />
                  <!-- AG Row Container bottomFullWidth -->
                  <div
                    class="ag-floating-bottom-full-width-container"
                    role="presentation"
                  />
                </div>
                <!-- AG Fake Horizontal Scroll -->
                <div
                  aria-hidden="true"
                  class="ag-body-horizontal-scroll ag-hidden"
                  style="height: 0px; max-height: 0px; min-height: 0px;"
                >
                  
            
                  <div
                    class="ag-horizontal-left-spacer"
                    ref="eLeftSpacer"
                    style="width: 0px; max-width: 0px; min-width: 0px;"
                  />
                  
            
                  <div
                    class="ag-body-horizontal-scroll-viewport"
                    ref="eViewport"
                    style="height: 0px; max-height: 0px; min-height: 0px;"
                  >
                    
                
                    <div
                      class="ag-body-horizontal-scroll-container"
                      ref="eContainer"
                      style="height: 0px; max-height: 0px; min-height: 0px; width: 0px;"
                    />
                    
            
                  </div>
                  
            
                  <div
                    class="ag-horizontal-right-spacer"
                    ref="eRightSpacer"
                    style="width: 0px; max-width: 0px; min-width: 0px;"
                  />
                  
        
                </div>
                <!-- AG Overlay Wrapper -->
                <div
                  aria-hidden="true"
                  class="ag-overlay ag-hidden"
                >
                  
            
                  <div
                    class="ag-overlay-panel"
                  >
                    
                
                    <div
                      class="ag-overlay-wrapper ag-layout-normal"
                      ref="eOverlayWrapper"
                    />
                    
            
                  </div>
                  
        
                </div>
              </div>
              <div
                class="ag-tab-guard ag-tab-guard-bottom"
                role="presentation"
                tabindex="0"
              />
            </div>
            <div
              class="ag-paging-panel ag-unselectable ag-hidden"
              id="ag-2"
            >
              
                
              <span
                class="ag-paging-row-summary-panel"
                role="status"
              >
                
                    
                <span
                  class="ag-paging-row-summary-panel-number"
                  id="ag-2-first-row"
                  ref="lbFirstRowOnPage"
                />
                
                    
                <span
                  id="ag-2-to"
                >
                  to
                </span>
                
                    
                <span
                  class="ag-paging-row-summary-panel-number"
                  id="ag-2-last-row"
                  ref="lbLastRowOnPage"
                />
                
                    
                <span
                  id="ag-2-of"
                >
                  of
                </span>
                
                    
                <span
                  class="ag-paging-row-summary-panel-number"
                  id="ag-2-row-count"
                  ref="lbRecordCount"
                />
                
                
              </span>
              
                
              <span
                class="ag-paging-page-summary-panel"
                role="presentation"
              >
                
                    
                <div
                  aria-label="First Page"
                  class="ag-paging-button"
                  ref="btFirst"
                  role="button"
                >
                  <span
                    class="ag-icon ag-icon-first"
                    role="presentation"
                    unselectable="on"
                  />
                </div>
                
                    
                <div
                  aria-label="Previous Page"
                  class="ag-paging-button"
                  ref="btPrevious"
                  role="button"
                >
                  <span
                    class="ag-icon ag-icon-previous"
                    role="presentation"
                    unselectable="on"
                  />
                </div>
                
                    
                <span
                  class="ag-paging-description"
                  role="status"
                >
                  
                        
                  <span
                    id="ag-2-start-page"
                  >
                    Page
                  </span>
                  
                        
                  <span
                    class="ag-paging-number"
                    id="ag-2-start-page-number"
                    ref="lbCurrent"
                  />
                  
                        
                  <span
                    id="ag-2-of-page"
                  >
                    of
                  </span>
                  
                        
                  <span
                    class="ag-paging-number"
                    id="ag-2-of-page-number"
                    ref="lbTotal"
                  />
                  
                    
                </span>
                
                    
                <div
                  aria-label="Next Page"
                  class="ag-paging-button"
                  ref="btNext"
                  role="button"
                >
                  <span
                    class="ag-icon ag-icon-next"
                    role="presentation"
                    unselectable="on"
                  />
                </div>
                
                    
                <div
                  aria-label="Last Page"
                  class="ag-paging-button"
                  ref="btLast"
                  role="button"
                >
                  <span
                    class="ag-icon ag-icon-last"
                    role="presentation"
                    unselectable="on"
                  />
                </div>
                
                
              </span>
              
            
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
